### ISSUE TITLE
allow repository_rules to run on stale state

### ISSUE THREAD

**Issue by dgoldstein0** [bazelbuild/bazel#23720]
### Description of the feature request:

repository_rules are a way to run arbitrary imperative operations during bazel's loading phase.  The default is that bazel clears their output folder before each execution of the repository rule - a good default, but it makes some use cases challenging.   I propose adding a parameter to repository_rule: `clear_output_folder_before_fetch` which should be True by default (maintaining current behavior), but could be set to False for use cases that trust their rules to properly handle stale state.

### Which category does this issue belong to?

Rules API

### What underlying problem are you trying to solve with this feature?

In particular, I've written a repository rule using yarn to install node_modules/ which we've integrated with our custom js library rules.  Yarn itself has a ton of logic to handle incremental updating of the generated node_modules/, so that `yarn install` after a small change to the yarn.lock can take a few seconds rather than a minute or more.  Since bazel clears the repository rule output folder, either we end up with quite bad performance, or we need quite a bit of extra complexity to put the outputs in a different folder and symlink them to bazel's output folder for the repository rule, so that we only have to create and rewrite symlinks on each repeat invocation and let yarn take care of the rest, as opposed to starting from scratch.

Likely other "use a package manager or other build system in a repository rule" use cases have this problem as well.

### Which operating system are you running Bazel on?

linux - ubuntu 22.04

### What is the output of `bazel info release`?

release 6.1.0-1c2df03733503215490047c398d25fe5b5553a0e

**Comment by mwoehlke-kitware** [comment]
> Likely other "use a package manager or other build system in a repository rule" use cases have this problem as well.

Can confirm; ran into this same problem with a repository that creates a Python virtual environment.

Also, repository rules sometimes run even when there has been *no* change. Why? (In particular, `bazel shutdown` triggers a re-run.)

**Comment by Wyverald** [comment]
> Also, repository rules sometimes run even when there has been _no_ change. Why? (In particular, `bazel shutdown` triggers a re-run.)

This sounds like a bug. If you could pin it down in a minimal repro case and file an issue, we'd appreciate it!

**Comment by jwnimmer-tri** [comment]
In @mwoehlke-kitware's case we were purposefully setting `local = True` for other reasons, in which case I believe re-running after a shutdown is the desired / specified behavior and there is no bug.

**Comment by meteorcloudy** [comment]
Sounds like interesting idea, @dgoldstein0 if you want to contribute, a simple design doc would be nice!

**Comment by dgoldstein0** [comment]
I'm pretty busy but I can try to squeeze it in.  is there a template or process I should use?

**Comment by meteorcloudy** [comment]
Yes, please follow https://bazel.build/contribute/design-documents
